lib/commons/dom/create-grid.js

```diff
@@ -248,50 +253,68 @@ function isFlexOrGridContainer(vNode) {
-  // @see https://www.w3.org/TR/css-flexbox-1/#painting
-  // @see https://www.w3.org/TR/css-grid-1/#z-order
-  if (
-    !['auto', '0'].includes(zIndex) &&
-    (positioned || isFlexOrGridContainer(parentVNode))
-  ) {
-    // if a positioned element has a z-index > 0, find the first
-    // true stack (not a "fake" stack created from positioned or
-    // floated elements without a z-index) and create a new stack at
-    // that point (step #5 and step #8)
-    // @see https://www.w3.org/Style/css2-updates/css2/zindex.html
-    while (stackingOrder.find(value => value % 1 !== 0)) {
-      const index = stackingOrder.findIndex(value => value % 1 !== 0);
-      stackingOrder.splice(index, 1);
-    }
-    stackingOrder[stackingOrder.length - 1] = parseInt(zIndex);
-  }
-  if (isStackingContext(vNode, parentVNode)) {
-    stackingOrder.push(0);
-  }
   // if a positioned element has z-index: auto or 0 (step #8), or if
   // a non-positioned floating element (step #5), treat it as its
   // own stacking context
   // @see https://www.w3.org/Style/css2-updates/css2/zindex.html
-  else if (positioned) {
-    // Put positioned elements above floated elements
-    stackingOrder.push(0.5);
-  } else if (floated) {
-    // Put floated elements above z-index: 0
-    // (step #5 floating get sorted below step #8 positioned)
-    stackingOrder.push(0.25);
+  if (!isStackingContext(vNode, parentVNode)) {
+    if (vNode.getComputedStylePropertyValue('position') !== 'static') {
+      // Put positioned elements above floated elements
+      stackingOrder.push(POSITION_STATIC_ORDER);
+    } else if (vNode.getComputedStylePropertyValue('float') !== 'none') {
+      // Put floated elements above z-index: 0
+      // (step #5 floating get sorted below step #8 positioned)
+      stackingOrder.push(FLOAT_ORDER);
+    }
+    return stackingOrder;
   }
 
+  // if an element creates a stacking context, find the first
+  // true stack (not a "fake" stack created from positioned or
+  // floated elements without a z-index) and create a new stack at
+  // that point (step #5 and step #8)
+  // @see https://www.w3.org/Style/css2-updates/css2/zindex.html
+  const index = stackingOrder.findIndex(value =>
+    [ROOT_ORDER, FLOAT_ORDER, POSITION_STATIC_ORDER].includes(value)
+  );
+  if (index !== -1) {
+    stackingOrder.splice(index, stackingOrder.length - index);
+  }
+
+  const zIndex = getRealZIndex(vNode, parentVNode);
+  if (!['auto', '0'].includes(zIndex)) {
+    stackingOrder.push(parseInt(zIndex));
+    return stackingOrder;
+  }
+  // since many things can create a new stacking context without position or
+  // z-index, we need to know the order in the dom to sort them by. Use the
+  // nodeIndex property to create a number less than the "fake" stacks from
+  // positioned or floated elements but still larger than 0
+  // 10 pad gives us the ability to sort up to 1B nodes (padStart does not
+  // exist in ie11)
+  let float = vNode.nodeIndex.toString();
```

Comment:
Can we rely on this? I was under the impression that we were always building the entire tree, but since you say that isn't the case, wouldn't we run into problems here if we were to compare a node that's in the tree with one that isn't?

Reply:
Changed it to use a nodeIndex from looping over the treewalker, which is guaranteed to get everything in DOM order.